Cython/Compiler/ExprNodes.py

```diff
@@ -2456,12 +2457,16 @@ def calculate_result_code(self):
             return "<error>"  # There was an error earlier
         if self.entry.is_cpp_optional and not self.is_target:
             return "(*%s)" % entry.cname
-        return entry.cname
+        return self.module_state_lookup + entry.cname
 
     def generate_result_code(self, code):
         entry = self.entry
         if entry is None:
             return  # There was an error earlier
+        if entry.scope.is_module_scope and (
+                entry.is_pyglobal or entry.is_cclass_var_entry):
+            # TODO - eventually this should apply to cglobals too
+            self.module_state_lookup = code.name_in_module_state("")
```

Comment:
Assigning this in `generate_result_code()` basically wormholes state between methods, which is always surprising.

The decision only seems to be related to the `entry`, so we should be able to tell at some point during the analysis phase.

Reply:
> The decision only seems to be related to the `entry`, so we should be able to tell at some point during the analysis phase.

Not sure that's right - it depends on the scope that it's called from rather than the entry. Although I agree it can be moved to the analysis phase.